dev/test/bulk-writer.ts

```diff
@@ -0,0 +1,535 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {expect} from 'chai';
+import {Status} from 'google-gax';
+
+import {Firestore, setLogFunction, Timestamp, WriteResult} from '../src';
+import {BulkWriter} from '../src/bulk-writer';
+import {Deferred} from '../src/util';
+import {
+  ApiOverride,
+  createInstance,
+  response,
+  verifyInstance,
+} from './util/helpers';
+
+// Change the argument to 'console.log' to enable debug output.
+setLogFunction(() => {});
+
+const PROJECT_ID = 'test-project';
+
+interface BulkWriterRequest {
+  writes: Array<{[key: string]: {} | undefined}>;
+}
+
+interface BulkWriterResponse {
+  writeResults: Array<{[key: string]: {} | null}>;
+  status: Array<{[key: string]: {}}>;
+}
+
+interface RequestResponse {
+  request: BulkWriterRequest;
+  response: BulkWriterResponse;
+}
+
+describe.only('BulkWriter', () => {
+  const documentNameRoot = `projects/${PROJECT_ID}/databases/(default)/documents/`;
+
+  let firestore: Firestore;
+  let bulkWriter: BulkWriter;
+  let requestCounter: number;
+  let opCount: number;
+  const flushDeferred = new Deferred<void>();
+  let flushCounter = 0;
+
+  beforeEach(() => {
+    requestCounter = 0;
+    opCount = 0;
+  });
+
+  function incrementOpCount(): void {
+    opCount++;
+  }
+
+  function verifyOpCount(expected: number): void {
+    expect(opCount).to.equal(expected);
+  }
+
+  function setOp(doc: string, value: string): {[key: string]: {}} {
```

Comment:
I tried rewriting to use the helpers. Creating a request that includes multiple operations is a bit hacky since we have to concat all the individual writes from all the created requests. The `update` helper also fills a precondition if none is provided, so I had to manually delete `currentDocument`. If you prefer this layout, I can update the tests in a separate PR to pass in a precondition instead of having it autofill.